### PR TITLE
Remove unnecessary device_guard usage

### PR DESCRIPTION
**Pull request by upsj**
pulled out of #938, we don't need to set the device ID kernels, because it already gets set by the executor->run function.

### CODE REVIEW

**Review by yhmtsai — APPROVED** [review]

File: cuda/factorization/ilu_kernels.cu [hunk 1]
```diff
@@ -57,7 +56,6 @@ void compute_lu(std::shared_ptr<const DefaultExecutor> exec,
 {
     const auto id = exec->get_device_id();
     auto handle = exec->get_cusparse_handle();
-    gko::cuda::device_guard g{id};
```
Comment (Slaedr) [inline cuda/factorization/ilu_kernels.cu:60]: Do you remember why this was only added for ic and ilu, but not other preconditioners?

**Review by Slaedr — APPROVED** [review]
LGTM! Thanks for the quick PR. However, why do we need to set the device at every kernel run? Can't we just set it once when the executor is created? I think we don't currently set it in executor constructors, though the device ID is already available at that point.

File: cuda/factorization/ilu_kernels.cu (hunk 1 above)
Comment (upsj) [reply cuda/factorization/ilu_kernels.cu:60]: no, I must have not remembered that Executor already sets them.

**Comment by upsj** [comment]
That's the issue with global state - we have no way of knowing whether it was modified somewhere, so in the worst case users change the device ID inside a custom LinOp::apply and we run solver kernels on GPU 1 while the data resides on GPU 0, which gives a hard to identify access violation.

**Comment by Slaedr** [comment]
I see. Okay, in general, that makes sense.

File: cuda/base/device_guard.hpp [hunk 2]
```diff
@@ -60,7 +60,9 @@ class device_guard {
     device_guard(int device_id)
     {
         GKO_ASSERT_NO_CUDA_ERRORS(cudaGetDevice(&original_device_id));
-        GKO_ASSERT_NO_CUDA_ERRORS(cudaSetDevice(device_id));
+        if (original_device_id != device_id) {
```
Comment (pratikvn) [inline cuda/base/device_guard.hpp:63]: :+1:

**Review by pratikvn — APPROVED** [review]
LGTM!

File: cuda/base/device_guard.hpp (hunk 2 above)
Comment (Slaedr) [inline cuda/base/device_guard.hpp:63]: Nice.

**Review by Slaedr — APPROVED** [review]
LGTM. That's even better!

**Review by tcojean — APPROVED** [review]

File: cuda/base/device_guard.hpp [hunk 3]
```diff
@@ -57,10 +57,13 @@ namespace cuda {
  */
 class device_guard {
 public:
-    device_guard(int device_id)
+    device_guard(int device_id) : original_device_id{}, need_reset{}
```
Comment (pratikvn) [inline cuda/base/device_guard.hpp:60]: For readability, might be better to be explicit and initialize it to `false` ? 
```suggestion
    device_guard(int device_id) : original_device_id{}, need_reset{false}
```

File: hip/base/device_guard.hip.hpp [hunk 4]
```diff
@@ -57,10 +57,13 @@ namespace hip {
  */
 class device_guard {
 public:
-    device_guard(int device_id)
+    device_guard(int device_id) : original_device_id{}, need_reset{}
```
Comment (upsj) [inline hip/base/device_guard.hip.hpp:60]:
```suggestion
    device_guard(int device_id) : original_device_id{}, need_reset{false}
```

**Pull request merged**